skimage/restoration/non_local_means.py

```diff
@@ -132,16 +139,34 @@ def denoise_nl_means(image, patch_size=7, patch_distance=11, h=0.1,
     if image.ndim != 3:
         raise NotImplementedError("Non-local means denoising is only \
         implemented for 2D grayscale and RGB images or 3-D grayscale images.")
+
+    if preserve_range is None and np.issubdtype(image.dtype, np.integer):
+        warn('Image dtype is not float. By default denoise_nl_means will '
+             'assume you want to preserve the range of your image '
+             '(preserve_range=True). In scikit-image 0.19 this behavior will '
+             'change to preserve_range=False. To avoid this warning, '
+             'explicitly specify the preserve_range parameter.',
+             stacklevel=2)
+        preserve_range = True
+
+    image = convert_to_float(image, preserve_range)
+    h = image.dtype.type(h)
+    sigma = image.dtype.type(sigma)
+
+    ctype = 'float32_t' if image.dtype == np.float32 else 'float64_t'
```

Comment:
This needs to use `float64_t` if `fast_mode` is True to avoid the numerical precision issues fixed previously in #2878. If the input was float32 for fast_mode we could still convert back to float32 after the Cython call to _nl_means_denoising..., but should use float64 internally.

Reply:
Hi @grlee77, thank you very much for  pointing me out this very interesting issue, I will definitely give a look to the provided reference there!
In this PR, the input image is provided to `convert_to_float`, and any array with a dtype different then `float32` will become `float64`.
The use of `float32` arrays is generally intentional, and I think that users doing so are aware of this kind of problem :-)
Using this strategy gives the user full control on what he wants to do: if he cares about the precision issues, he only has to cast the input array to double by himself...

Reply:
> The use of float32 arrays is generally intentional, and I think that users doing so are aware of this kind of problem :-)

I doubt most users will be aware of the potential issue in this particular case. It is not just an issue of precision down at the level of 1e-7 or so, but potentially highly degraded output. A compromise could be to just raise a warning on float32 input with fast_mode True.

That PR discussion is a bit long, but there is a [nice visual example in this comment](https://github.com/scikit-image/scikit-image/pull/2878#issuecomment-346344773). Basically there are cumulative sums being used that become increasingly inaccurate for larger image sizes.

Reply:
> I doubt most users will be aware of the potential issue in this particular case.

You're right... I just had a look at *fast_exp.h* and at the reference paper, the implemented `fast_exp` is double precision specific, and it is not really clear to me what cython is doing for single precision inputs...

I will investigate and see if there is any single precision alternative...

Reply:
According to your [answer](https://github.com/scikit-image/scikit-image/pull/2878#issuecomment-346176122) 
>  I see that you have changed both variants to use double precision, but I think only the fast, integral image approach will have a precision that decreases with increasing image size.

Isn't only arrays with large integral values that suffers for the approximation error?
Using `convert_to_float` this kind of arrays is converted to `float64`...
Now the open question is what about `float32` with large values? Should some tests asserting `fast_exp` behavior with float32 arrays be added?

Reply:
Yes, the arrays that needs to always remain double precision is `integral` as used in `_integral_image_2d` and `_integral_image_3d`. This is where the large size of the summations accumulates a lot of error for large images/volumes. 

It looks like this is indeed still the case here so +1 from me